cc/dual_net/dual_net.h

```diff
@@ -51,6 +55,42 @@ class DualNet {
   // TODO(tommadams): Change features type from float to uint8_t.
   using StoneFeatures = std::array<float, kNumStoneFeatures>;
   using BoardFeatures = std::array<float, kNumBoardFeatures>;
+  using Policy = std::array<float, kNumMoves>;
+
+  struct Result {
+    std::vector<Policy> policies;
+    std::vector<float> values;
+    std::string model;
+  };
+
+  class Service;
+
+  // Base class to perform inferences on a single batch of features.
+  class Client {
+   public:
+    virtual ~Client();
+
+    // Runs inference on a batch of input features. Thread-safe.
+    virtual Result RunMany(std::vector<BoardFeatures>&& features) = 0;
```

Comment:
Thanks for the refactor, this helped me understand the code structure better. I still don't think the Client class is required and batching would be better implemented as a subclass of DualNet. As far as I can tell DualNet::Service is basically the same as DualNetFactory (if batching were implemented as a DualNetFactory), so I don't think that is required either.

Both DualNet::RunMany and DualNet::Client::RunMany have very similar interfaces: they both operate on batches of games, the difference is just that the DualNet::RunMany takes flattened inputs. The DualNet implementations are going to have to flatten those games into a single inference batch anyway.

The way I see it, batching should be purely an implementation detail of a simple DualNet subclass that wraps a concrete DualNet (e.g. TfDualNet). In the example I gave in https://github.com/tommadams/minigo/commit/291abff7061e88ab1039d9d080a4b42db834f9c7, the BatchingDualNet wrapped the concrete DualNet implementation.

I think this is close to what you have already and we can probably get there with a little more iteration.

Reply:
One of the reasons I'm pushing back quite strongly against adding complexity to this part of the codebase is that we're in the process of porting the C++ code to Raspberry Pi and we hope that will encourage people who are interested in Go and ML to play around with the code. So we're placing a higher than usual emphasis on code simplicity, especially around the inference code where we'd like people to be able to easily drop in a new implementation.

(The existing code isn't as simple as I'd like FWIW)